ibm/resource_ibm_kms_key_rings.go

```diff
@@ -0,0 +1,248 @@
+
+	var hpcsEndpointURL string
+
+	if crnData[4] == "hs-crypto" {
+		hpcsEndpointAPI, err := meta.(ClientSession).HpcsEndpointAPI()
+		if err != nil {
+			return err
+		}
+
+		resp, err := hpcsEndpointAPI.Endpoint().GetAPIEndpoint(instanceID)
+		if err != nil {
+			return err
+		}
+
+		if endpointType == "public" {
+			hpcsEndpointURL = "https://" + resp.Kms.Public + "/api/v2/keys"
+		} else {
+			hpcsEndpointURL = "https://" + resp.Kms.Private + "/api/v2/keys"
+		}
+
+		u, err := url.Parse(hpcsEndpointURL)
+		if err != nil {
+			return fmt.Errorf("Error Parsing hpcs EndpointURL")
+		}
+		kpAPI.URL = u
+	} else if crnData[4] == "kms" {
+		if endpointType == "private" {
+			if !strings.HasPrefix(kpAPI.Config.BaseURL, "private") {
+				kpAPI.Config.BaseURL = "private." + kpAPI.Config.BaseURL
+			}
+		}
+	} else {
+		return fmt.Errorf("Invalid or unsupported service Instance")
+	}
+	kpAPI.Config.InstanceID = instanceID
+
+	err = kpAPI.CreateKeyRing(context.Background(), keyRingID)
+	if err != nil {
+		return fmt.Errorf(
+			"Error while creating key ring : %s", err)
+	}
+	var keyRing string
+	keyRings, err2 := kpAPI.GetKeyRings(context.Background())
+	if err2 != nil {
+		return fmt.Errorf(
+			"Error while fetching key ring : %s", err2)
+	}
+	for _, v := range keyRings.KeyRings {
+		log.Print(v)
+		if v.ID == d.Get("key_ring_id") {
+			keyRing = v.ID
+		}
+	}
+
+	d.SetId(fmt.Sprintf("%s:keyRing:%s", keyRing, instanceData.Crn.String()))
+
+	return resourceIBMKmsKeyRingRead(d, meta)
+}
+
+func resourceIBMKmsKeyRingRead(d *schema.ResourceData, meta interface{}) error {
+	kpAPI, err := meta.(ClientSession).keyManagementAPI()
+	if err != nil {
+		return err
+	}
+	id := strings.Split(d.Id(), ":keyRing:")
+	crn := id[1]
+	crnData := strings.Split(crn, ":")
+	endpointType := crnData[3]
+	instanceID := crnData[len(crnData)-3]
+
+	var hpcsEndpointURL string
+
+	if crnData[4] == "hs-crypto" {
+		hpcsEndpointAPI, err := meta.(ClientSession).HpcsEndpointAPI()
+		if err != nil {
+			return err
+		}
+
+		resp, err := hpcsEndpointAPI.Endpoint().GetAPIEndpoint(instanceID)
+		if err != nil {
+			return err
+		}
+
+		if endpointType == "public" {
+			hpcsEndpointURL = "https://" + resp.Kms.Public + "/api/v2/keys"
+		} else {
+			hpcsEndpointURL = "https://" + resp.Kms.Private + "/api/v2/keys"
+		}
+
+		u, err := url.Parse(hpcsEndpointURL)
+		if err != nil {
+			return fmt.Errorf("Error Parsing hpcs EndpointURL")
+
+		}
+		kpAPI.URL = u
+	} else if crnData[4] == "kms" {
+		if endpointType == "private" {
+			if !strings.HasPrefix(kpAPI.Config.BaseURL, "private") {
+				kpAPI.Config.BaseURL = "private." + kpAPI.Config.BaseURL
+			}
+		}
+	} else {
+		return fmt.Errorf("Invalid or unsupported service Instance")
+	}
+
+	kpAPI.Config.InstanceID = instanceID
+	_, err = kpAPI.GetKeyRings(context.Background())
+	if err != nil {
+		kpError := err.(*kp.Error)
+		if kpError.StatusCode == 404 || kpError.StatusCode == 409 {
+			return nil
+		} else {
```

Comment:
no need of else block 
```
else {
			return fmt.Errorf("Get Key Rings failed with error: %s", err)
		}
```
instead you can directly do `return fmt.Errorf("Get Key Rings failed with error: %s", err)`

Reply:
Need Discussion